### PR TITLE
Support for u0 initial guess function

### PR DESCRIPTION
**Pull request by avik-pal**
Fixes https://github.com/SciML/BoundaryValueDiffEq.jl/issues/128

## TODOs

- [x] Upstream `get_concrete_u0` patch so that `u0` as a function is propagated unlike IVPs
- [x] Add tests (use the one in https://github.com/SciML/BoundaryValueDiffEq.jl/issues/117#issuecomment-1780981510)

### CODE REVIEW

**Comment by avik-pal** [comment]
I will fix the piracy once https://github.com/SciML/DiffEqBase.jl/pull/959 is merged and released.

**Comment by avik-pal** [comment]
I want to get this in before https://github.com/SciML/BoundaryValueDiffEq.jl/pull/131, since that one is major release.

File: src/utils.jl [hunk 1]
```diff
@@ -139,17 +139,38 @@ function __extract_problem_details(prob, u0::AbstractArray; dt = 0.0,
...
+    u0 = __initial_guess(f, prob.p, prob.tspan[1])
+    t₀, t₁ = prob.tspan
+    return Val(true), eltype(u0), length(u0), Int(cld(t₁ - t₀, dt)), u0
 end
 
-__initial_state_from_prob(prob::BVProblem, mesh) = __initial_state_from_prob(prob.u0, mesh)
-__initial_state_from_prob(u0::AbstractArray, mesh) = [copy(vec(u0)) for _ in mesh]
-function __initial_state_from_prob(u0::AbstractVector{<:AbstractVector}, _)
+function __initial_guess(f::F, p::P, t::T) where {F, P, T}
+    if static_hasmethod(f, Tuple{P, T})
```
Comment (ChrisRackauckas) [inline src/utils.jl:152]: note that in v1.10 this isn't needed, and it can be incorrect.

File: src/utils.jl [hunk 2]
```diff
@@ -139,17 +139,38 @@ function __extract_problem_details(prob, u0::AbstractArray; dt = 0.0,
...
+        return f(p, t)
+    elseif static_hasmethod(f, Tuple{T})
+        return f(t)
+    else
+        throw(ArgumentError("`initial_guess` must be a function of the form `f(p, t)` or \
+                             `f(t)`"))
```
Comment (ChrisRackauckas) [inline src/utils.jl:158]: We should just simplify it to be just one form.

Comment (avik-pal) [reply src/utils.jl:158]: We can only allow `f(t)` according to the SciMLBase specification but I think we almost need the parameters

Comment (avik-pal) [reply src/utils.jl:158]: Ideally we just support `f(p, t)`. But that is a breaking change following the SciMLBase docs

Comment (ChrisRackauckas) [reply src/utils.jl:158]: throw an warning that f(t) is deprecated. Might as well start pushing in the right direction.

Comment (avik-pal) [reply src/utils.jl:158]: So let it work right now with the `static_hasmethod` but just throw a warning, correct?

Comment (ChrisRackauckas) [reply src/utils.jl:158]: yes

Comment (avik-pal) [reply src/utils.jl]: Done

**Pull request merged**